guillotina_amqp/amqp.py

```diff
@@ -84,6 +96,137 @@
+        exchange_name='beacon',
+        queue_name=beacon_queue_name,
+        routing_key=beacon_queue_name,
+    )
+
+    # Declare and bind delay beacon temporary queue
+    await channel.queue_declare(
+        queue_name=beacon_delay_queue_name,
+        exclusive=True,  # This deletes the queue if connection is lost
+        arguments={
+            'x-dead-letter-exchange': 'beacon',
+            'x-dead-letter-routing-key': beacon_queue_name,
+            'x-message-ttl': ttl,  # 30 seconds
+        })
+
+    await channel.queue_bind(
+        exchange_name='beacon',
+        queue_name=beacon_delay_queue_name,
+        routing_key=beacon_delay_queue_name,
+    )
+
+    # Configure main beacon message handler
+    await channel.basic_consume(
+        handle_beacon,
+        queue_name=beacon_queue_name,
+    )
+
+    # Start the autokill task
+    autokill_handler = asyncio.ensure_future(autokill(ttl))
+
+    # Send the first beacon
+    asyncio.ensure_future(publish_beacon_to_delay_queue(channel, ttl))
+
+
+async def autokill(ttl):
+    """
+    AutoKill task, sleeps as long as beacons are received, and exits the
+    process otherwise
+
+    This task captures cancel events (cancel is called from the beacon handler)
+    and keeps sleeping in that case, otherwise exits the process because that
+    means we didn't receive any beacons (AMQP connection is stalled)
+    """
+    loop = asyncio.get_event_loop()
+
+    while True:
+        try:
+            asleep = ttl * 3
+            logger.debug(f'autokill: sleeping for {asleep} seconds')
+            await asyncio.sleep(asleep)
+        except asyncio.CancelledError:
+            logger.debug(f'autokill: received cancel, beacon was received')
+            continue
+        except Exception as err:
+            logger.debug(f'autokill: unknown exception {err}')
+            continue
+        else:
+            logger.debug(f'autokill: No beacons received')
+            break
+
+    # No beacons received: set the autokill event, and schedule a call
+    # to os._exit a bit later so that unit tests can wait on autokill_event
+
+    logger.error(f'Exiting worker because of no beacon activity')
+
+    autokill_event.set()
+    loop.call_later(2, os._exit, 0)
+
+
+async def publish_beacon_to_delay_queue(channel, wait=0):
+    # Publish to beacon delay queue
+    await asyncio.sleep(wait)
+    try:
+        beacon_payload = json.dumps({'worker_beacon_uuid': worker_beacon_uuid})
+        await channel.publish(
+            beacon_payload,
+            exchange_name='beacon',
+            routing_key=beacon_delay_queue_name,
+            properties={
+                'delivery_mode': 2,
+            }
+        )
+    except aioamqp.exceptions.ChannelClosed:
+        logger.debug(f'Beacon publish: channel is closed !')
+    else:
+        logger.debug(f'Beacon published: {beacon_payload}')
+
+
+async def handle_beacon(channel, body, envelope, properties):
+    global autokill_handler
```

Comment:
Maybe instead of using globals we could use a class and have these of member properties? Then `handle_beacon` is a instance method. globals are more okay in asyncio but maybe we could do without them?

Reply:
We killed the globals with PR #11 :)